libs/gltfio/src/ResourceLoader.cpp

```diff
@@ -172,24 +167,26 @@ static void importSkinningData(Skin& dstSkin, const cgltf_skin& srcSkin) {
     }
 }
 
-void ResourceLoader::addResourceData(const char* uri, BufferDescriptor&& buffer) {
-    pImpl->mUriDataCache.emplace(uri, std::move(buffer));
-}
-
-bool ResourceLoader::hasResourceData(const char* uri) const {
-    return pImpl->mUriDataCache.find(uri) != pImpl->mUriDataCache.end();
-}
-
 static void convertBytesToShorts(uint16_t* dst, const uint8_t* src, size_t count) {
     for (size_t i = 0; i < count; ++i) {
         dst[i] = src[i];
     }
 }
 
-static void generateTrivialIndices(uint32_t* dst, size_t numVertices) {
-    for (size_t i = 0; i < numVertices; ++i) {
-        dst[i] = i;
-    }
+ResourceLoader::ResourceLoader(const ResourceConfiguration& config) :
+        mPool(new AssetPool), pImpl(new Impl(config)) { }
+
+ResourceLoader::~ResourceLoader() {
+    mPool->onLoaderDestroyed();
```

Comment:
I *really* hate callbacks from destructors -- this is because dtors can be called from weird places, in particular places that hold locks. It also means the object calling the callback is half-destroyed, so of the callback has a way to get to a pointer to it, it can be catastrophic (especially if they can take shared_ptr to it).
Not saying this is the case here, just that in general, this is something to avoid.

That's why I prefer having a `terminate` method on objects that have side effects at destruction time.

At the very least, the documentation needs to state very clearly that `onLoaderDestroyed` is called from the dtor of `ResourceLoader` and state whether it is safe to take locks from the callback (and if so, that becomes part of the API).

Reply:
`onLoaderDestroyed` is perhaps badly named since this isn't an actual callback, and it certainly doesn't result in any user-defined functions being called.  It simply tells the AssetPool to release a shared reference.  In fact I could implement this using `shared_ptr` instead, it might be more readable that way.

I generally agree with your advice regarding callbacks within destructors, although I think they have served us well in `BufferDescriptor`.